### PR TITLE
Divide and conquer traverse 2

### PR DESCRIPTION
**Pull request by jacereda**
This is https://github.com/purescript/purescript-foldable-traversable/pull/54 relative to current master.

### CODE REVIEW

**Comment by paf31** [comment]
@S11001001 @hdgarrood Does this look okay to everyone now? I assume it's been reviewed several times already 😄

File: test/Main.purs [hunk 1]
```diff
@@ -32,9 +32,9 @@ foldableLength = unwrap <<< foldMap (const (Additive 1))
 deferEff :: forall e a. (Unit -> a) -> Eff e a
 deferEff = unsafeCoerce
 
-type EffTest = Eff (console :: CONSOLE, assert :: ASSERT)
+type EffTest a = forall e. Eff (assert :: ASSERT | e) a
 
-main :: EffTest Unit
```
Comment (hdgarrood) [inline test/Main.purs]: Could you please leave this as it was?

File: test/Main.purs [hunk 2]
```diff
@@ -32,9 +32,9 @@ foldableLength = unwrap <<< foldMap (const (Additive 1))
 deferEff :: forall e a. (Unit -> a) -> Eff e a
 deferEff = unsafeCoerce
 
-type EffTest = Eff (console :: CONSOLE, assert :: ASSERT)
+type EffTest a = forall e. Eff (assert :: ASSERT | e) a
```
Comment (hdgarrood) [inline test/Main.purs]: Could you explain the reason for this change?

File: test/Main.purs [hunk 3]
```diff
@@ -179,18 +184,77 @@ testFoldableFWith f n = do
...
+
+  assert $ foldrWithIndex (\i x y -> (i + 1) * x + y) 0 dat == expectedSum
+  assert $ foldlWithIndex (\i y x -> y + (i + 1) * x) 0 dat == expectedSum 
+  assert $ foldMapWithIndex (\i x -> Additive $ (i + 1) * x) dat == Additive expectedSum
+
+testFoldableWithIndexArrayWith :: Int -> EffTest Unit
+testFoldableWithIndexArrayWith = testFoldableWithIndexFWith arrayFrom1UpTo
+
+
+data Tuple a b = Tuple a b
```
Comment (hdgarrood) [inline test/Main.purs]: Surely we're already pulling this in as a dependency?

**Comment by hdgarrood** [comment]
Oh I see, they're from merging master. Ok then. :+1:

**Comment by jacereda** [comment]
The signatures aren't like that in master. Those were probably in conflict and maybe I resolved with that. I'll leave it as master but unifying `eff` vs `e`.

**Comment by garyb** [comment]
@jacereda, @S11001001 thanks for doing this, sorry it took so absurdly long to get merged!

**Pull request merged**